Compiler panic when wraping a divide-by-zero statement inside a function

<!--## Prerequisites

- First, many thanks for taking part in the community. We really appreciate that.
- Support questions are better asked on our [gitter channel](https://gitter.im/ZoKrates/ZoKrates)
- Ensure the issue isn't already reported.
- The issue should be reproducible with the latest ZoKrates version; however, this isn't a hard requirement and being reproducible with an older version is sufficient.

*Delete the above section and the instructions in the sections below before submitting*

-->

## Description

<!--Please shortly describe the bug you have found, and what you expect instead.-->

When wrapping a divide-by-zero assertion statement inside a function, the compiler crashes.

## Environment

- Compiler version: 0.8.7
- Operating system: Ubuntu 22.04

## Steps to Reproduce

Save the following file into `buggy_inline.zok`:

```zokrates
def foo(field a, field b) -> field {
    assert(1 / (a - b) == 0);
    return a;
}

def main() -> field {
    return foo(1, 1);
}
```

Compile:

```shell
zokrates compile -i buggy_inline.zok
```

and observe the following output:

```txt
The compiler unexpectedly panicked
panicked at 'called `Option::unwrap()` on a `None` value', zokrates_field/src/bn128.rs:4:1
This is unexpected, please submit a full bug report at https://github.com/Zokrates/ZoKrates/issues
make: *** [Makefile:8: out] Error 101
```

<!--
Please provide a *minimal* source code example to trigger the bug you have found.
Please also mention any command line flags that are necessary for triggering the bug.
Provide as much information as necessary to reproduce the bug.

```zokrates
// Some *minimal* ZoKrates source code to reproduce the bug.
// ...
```
-->

--- a/changelogs/unreleased/1326-dark64
+++ b/changelogs/unreleased/1326-dark64
@@ -1,0 +1,1 @@
+Detect division by zero on compile-time

--- a/zokrates_analysis/src/propagation.rs
+++ b/zokrates_analysis/src/propagation.rs
@@ -33,6 +33,7 @@ pub enum Error {
     InvalidValue(String),
     OutOfBounds(u128, u128),
     VariableLength(String),
+    DivisionByZero,
 }
 
 impl fmt::Display for Error {
@@ -47,6 +48,9 @@ impl fmt::Display for Error {
                 index, size
             ),
             Error::VariableLength(message) => write!(f, "{}", message),
+            Error::DivisionByZero => {
+                write!(f, "Division by zero detected during static analysis",)
+            }
         }
     }
 }
@@ -924,12 +928,16 @@ impl<'ast, T: Field> ResultFolder<'ast, T> for Propagator<'ast, T> {
                 let left = self.fold_field_expression(*e.left)?;
                 let right = self.fold_field_expression(*e.right)?;
 
-                Ok(match (left, right) {
-                    (FieldElementExpression::Value(n1), FieldElementExpression::Value(n2)) => {
-                        FieldElementExpression::Value(ValueExpression::new(n1.value / n2.value))
+                match (left, right) {
+                    (_, FieldElementExpression::Value(n)) if n.value == T::from(0) => {
+                        Err(Error::DivisionByZero)
                     }
-                    (e1, e2) => e1 / e2,
-                })
+                    (e, FieldElementExpression::Value(n)) if n.value == T::from(1) => Ok(e),
+                    (FieldElementExpression::Value(n1), FieldElementExpression::Value(n2)) => Ok(
+                        FieldElementExpression::Value(ValueExpression::new(n1.value / n2.value)),
+                    ),
+                    (e1, e2) => Ok(e1 / e2),
+                }
             }
             FieldElementExpression::Neg(e) => match self.fold_field_expression(*e.inner)? {
                 FieldElementExpression::Value(n) => {
@@ -1571,14 +1579,30 @@ mod tests {
 
             #[test]
             fn div() {
-                let e = FieldElementExpression::div(
-                    FieldElementExpression::value(Bn128Field::from(6)),
-                    FieldElementExpression::value(Bn128Field::from(2)),
+                let mut propagator = Propagator::default();
+
+                assert_eq!(
+                    propagator.fold_field_expression(FieldElementExpression::div(
+                        FieldElementExpression::value(Bn128Field::from(6)),
+                        FieldElementExpression::value(Bn128Field::from(2)),
+                    )),
+                    Ok(FieldElementExpression::value(Bn128Field::from(3)))
                 );
 
                 assert_eq!(
-                    Propagator::default().fold_field_expression(e),
-                    Ok(FieldElementExpression::value(Bn128Field::from(3)))
+                    propagator.fold_field_expression(FieldElementExpression::div(
+                        FieldElementExpression::identifier("a".into()),
+                        FieldElementExpression::value(Bn128Field::from(1)),
+                    )),
+                    Ok(FieldElementExpression::identifier("a".into()))
+                );
+
+                assert_eq!(
+                    propagator.fold_field_expression(FieldElementExpression::div(
+                        FieldElementExpression::value(Bn128Field::from(6)),
+                        FieldElementExpression::value(Bn128Field::from(0)),
+                    )),
+                    Err(Error::DivisionByZero)
                 );
             }
 

